### PR TITLE
Fixed disconnect; handle concurrent calls; fixed and updated tests

### PR DESCRIPTION
**Pull request by pulyaevskiy**
This fixes:

* `disconnect()` didn't actually do anything
* in my tests when I fixed `disconnect()` it's response was empty `Map` which was failing in constructor of `GoogleSignInAccount`
* not sure if it's the best word, but fixed handling of "concurrent" calls to all methods, so that multiple calls to `signIn` for instance will only trigger sign in flow once. Plus it also ensures only single user update will be published in the stream.
* fixed and updated tests with a couple test cases.
* dartfmt

P.S.: sorry for yet another PR for disconnect issue.

### CODE REVIEW

**Comment by pulyaevskiy** [comment]
Now that I've spent some more time thinking about proposed fixes, there is a chance users can still experience strange behavior if, for instance:

1. user calls `signInSilently()`
2. without waiting for Future from (1) to complete user calls `signIn()`, which returns the same future.
3. In this case interactive sign in flow will not be trigger.

I think this is an unlikely edge case, but it should still be prevented in some way. The same applies to signOut/disconnect. Possible ways as I see:

1. Memoize each method separately
2. Add some assertions to prevent from calls which are not allowed

**Comment by pulyaevskiy** [comment]
Also, looks like build needs to be restarted:

```
BUILDING IPA for image_picker/example
Running "flutter packages get" in example...
HTTP error 500: Internal Server Error at https://pub.dartlang.org/api/packages/async.
This is likely a transient error. Please try again later.
pub get failed (1)
```

File: packages/google_sign_in/lib/google_sign_in.dart [hunk 1]
```diff
@@ -138,18 +134,42 @@ class GoogleSignIn {
...
 
   /// Disconnects the current user from the app and revokes previous
   /// authentication.
-  Future<GoogleSignInAccount> disconnect() => _callMethod('disconnect');
+  Future<GoogleSignInAccount> disconnect() => _signOutWithMethod('disconnect');
+
+  Future<GoogleSignInAccount> _signOutWithMethod(String method) {
+    if (_signOutCompleter != null) return _signOutCompleter.future;
+    _signOutCompleter = new Completer();
+    _signOutCompleter.complete(_callMethod(method).whenComplete(() {
```
Comment (goderbauer) [inline packages/google_sign_in/lib/google_sign_in.dart]: Instead of introducing a completer, can you not just hold on to the Future returned by _callMethod?

**Comment by goderbauer** [comment]
In the edge case where you call `signIn` after `signInSilently`, the ideal solution would probably be to check in the `signIn` call if a call to `signInSilently` is currently in progress, wait for the result and if it didn't return a valid user actually trigger the interactive sign in.

**Comment by pulyaevskiy** [comment]
Thanks @goderbauer !

To address both of your comments I think I can replace completers with a `Map<String, Future>` which holds currently active futures for each method. This would allow to implement a proper `signInSilently` followed by `signIn` flow.

Will try this out later today.

**Comment by pulyaevskiy** [comment]
Actually, since all the methods update `currentUser` it probably doesn't make sense to allow any of them called concurrently. A solution can be as simple as only keeping track of a Future returned from the most recent call to `_callMethod()` and preventing any other concurrent requests until it completes (I'm thinking simple `assert(_callFuture == null)` should be good enough?).

**Comment by collinjackson** [comment]
Looking at the native code, it seems like Android throws errors when it detects multiple concurrent requests on the native side, unless they are calls to the same method in which case they are merged. On iOS side, sign out requests are assumed to succeed, sign in requests are batched up and when the first one returns the result is used to respond to all of them.

From a developer's perspective, I think having the Dart side create the illusion of a deterministic API that doesn't throw unnecessary errors would be nice. It is tempting it call `signInSilently()` all over the place and expect it never complain at you. It even has "silent" in its name!

Right now there are several reasonable ways to get the user:

1. Throw away the result of signIn() and use `onCurrentUserChanged` with `StreamBuilder`
1. Retain the result of `googleSignIn.signIn` to and use a `FutureBuilder` to await it.
1. `await googleSignIn.signIn();` but then throw away the result and trigger a call `if(mounted) setState(() {});`. In the build method, read `currentUser` to get the latest data.

The first two are my favorite, but the proposed assertion would create a races, e.g. if the developer called signInSilently() in two places which is a reasonable thing to want to do.

The last is likely to be the most error-prone, but tempting for people who are building wrapper abstractions and don't want to use `FutureBuilder`/`StreamBuilder` directly.

I wonder if we should deprecate `currentUser` (or write a discouraging comment) and encourage people to `await googleSignIn.signInSilently()` if they need a user object.

**Comment by pulyaevskiy** [comment]
Thanks @collinjackson !

Deprecating `currentUser` would make sense to me. But it also seems like Android convention would work in our case too.

As long as concurrent calls to different methods are not allowed we should be able to guarantee valid state in `currentUser` if I'm not missing anything?

I've just pushed an update which implements Android version as described above. Looking forward for any feedback!

**Comment by collinjackson** [comment]
I think signInSilently / signIn are likely to be racing each other if an app includes several Flutter widgets use `FutureBuilder` or `StreamBuilder` to get a handle to the currently signed-in user. I would prefer that to work and not trigger assertions, rather than forcing developers to build their own wrapper code to serialize the calls across their widget tree. I think these patterns should work as they're not using `currentUser`.

What do you think? It would be straightforward to modify your _guardMethodCall could be modified to put the method calls on in a queue so that at most one is in flight at a time.

**Comment by pulyaevskiy** [comment]
Thanks! This makes sense to me. I refactored this PR to use something similar to what you suggested. Had to use a completer because it provides access to `isCompleted` and I wanted to prevent executing the same method more times than it needs to.

Also merged in latest master and resolved conflicts.
Haven't updated docs yet. If this looks good, I'll work on documentation next.

File: packages/google_sign_in/test/google_sign_in_test.dart [hunk 2]
```diff
@@ -2,103 +2,131 @@
...
-      stubResponse = _failureResponse();
 
-      Function invocationRecorder = (Invocation invocation) async {
-        invokedMethods.add(invocation.positionalArguments[0]);
-        return stubResponse;
-      };
-      // Mock with and without optional 2nd argument.
-      when(mockChannel.invokeMethod(any)).thenAnswer(invocationRecorder);
-      when(mockChannel.invokeMethod(any, any)).thenAnswer(invocationRecorder);
+      dynamic answer(Invocation invocation) {
```
Comment (collinjackson) [inline packages/google_sign_in/test/google_sign_in_test.dart]: Return type should be `Future` instead of `dynamic`

File: packages/google_sign_in/test/google_sign_in_test.dart [hunk 3]
```diff
@@ -2,103 +2,131 @@
...
+        final method = invocation.positionalArguments[0];
+        invokedMethods.add(method);
+        final response = responses[method];
+        return new Future.value(response);
+      }
 
-      googleSignIn = new GoogleSignIn.private(channel: mockChannel);
-    });
+      when(mockChannel.invokeMethod(typed(any), typed(any))).thenAnswer(answer);
```
Comment (collinjackson) [inline packages/google_sign_in/test/google_sign_in_test.dart]: Since these tests were written the `MethodChannel.setMockMethodCallHandler` method was added. I think we should use that so we can remove the dependency on package:mockito and rename `GoogleSignIn.private` to `GoogleSignIn._`. You can see this pattern used in the [shared_preferences](https://github.com/flutter/plugins/blob/master/packages/shared_preferences/test/shared_preferences_test.dart#L39) and [firebase_database](https://github.com/flutter/plugins/blob/master/packages/firebase_database/test/firebase_database_test.dart#L26) tests.

File: packages/google_sign_in/lib/google_sign_in.dart [hunk 4]
```diff
@@ -126,31 +124,63 @@ class GoogleSignIn {
...
+        ? new GoogleSignInAccount._(this, response)
+        : null;
     _streamController.add(_currentUser);
     return _currentUser;
   }
 
-  /// The currently signed in account, or null if the user is signed out
-  GoogleSignInAccount _currentUser;
+  /// Keeps track of the most recently scheduled method call.
+  Completer _lastMethodCompleter;
```
Comment (goderbauer) [inline packages/google_sign_in/lib/google_sign_in.dart]: I'd prefer to define a a super-small class with these two as final member variables. That makes it clearer that these two have to stay in sync.

File: packages/google_sign_in/lib/google_sign_in.dart [hunk 5]
```diff
@@ -126,31 +124,63 @@ class GoogleSignIn {
...
-  /// If there is already a signed-in user, returns it.
-  Future<GoogleSignInAccount> signInSilently() async => _currentUser ?? _callMethod('signInSilently');
+  Future<GoogleSignInAccount> signInSilently() =>
+      _addMethodCall('signInSilently');
 
   /// Starts the sign-in process.
-  ///
-  /// If there is already a signed-in user, returns it.
-  Future<GoogleSignInAccount> signIn() async => _currentUser ?? _callMethod('signIn');
+  Future<GoogleSignInAccount> signIn() => _addMethodCall('signIn');
```
Comment (goderbauer) [inline packages/google_sign_in/lib/google_sign_in.dart]: For my own understanding: This is changing the logic slightly, right? Previously, if you called `signInSilently`, waited for the user object, and then for whatever reason called `signIn` you got the same user object back. With this change, the call to `signIn` will actually be dispatched to the native side (will it bring up a new sign-in dialog there?). Is this change intended?

Comment (pulyaevskiy) [inline packages/google_sign_in/lib/google_sign_in.dart]: Good point. Yes, second call will trigger new sign-in flow on the native side.
I can probably make `_addMethodCall()` return currently signed in user for `signIn` and `signInSilently` if `currentUser != null`.

This should keep it consistent with original logic. How does this sound?

**Comment by pulyaevskiy** [comment]
Thanks @collinjackson and @goderbauer for your comments!

Please take a look at latest changes. I addressed all code related issues, I hope.
Attempted to add some documentation, but feel free to adjust/add more (not a native speaker here).

File: packages/google_sign_in/lib/google_sign_in.dart [hunk 6]
```diff
@@ -126,31 +119,80 @@ class GoogleSignIn {
...
+  Future<GoogleSignInAccount> _addMethodCall(String method) {
+    if (_lastMethodCompleter == null) {
+      _lastMethodCompleter = new _MethodCompleter(method);
+      _lastMethodCompleter.complete(_callMethod(method));
+      return _lastMethodCompleter.future;
+    }
+
+    // Return the same Future for consecutive calls to the same method.
+    if (_lastMethodCompleter.method == method &&
+        !_lastMethodCompleter.isCompleted) {
```
Comment (tvolkert) [inline packages/google_sign_in/lib/google_sign_in.dart]: Why not do the same thing if `isCompleted` is true?

Comment (pulyaevskiy) [inline packages/google_sign_in/lib/google_sign_in.dart]: We don't know if the last call completed successfully or with an error.
This only covers a situation when a second call to the same method comes while the first one is in-flight. Handling completed calls is done on lines 150-158.

This is actually more like a small optimization. Without this `if` block we'll get the same guarantees, the only difference is that each method call would always return a fresh Future instance.

So it should be safe to remove it, which maybe not a bad idea (less code, less troubles).

File: packages/google_sign_in/lib/google_sign_in.dart [hunk 7]
```diff
@@ -178,20 +220,37 @@ class GoogleUserCircleAvatar extends StatelessWidget {
...
 }
+
+class _MethodCompleter {
+  final String method;
+  final Completer<GoogleSignInAccount> _completer = new Completer();
+
+  _MethodCompleter(this.method);
+
+  void complete(FutureOr<GoogleSignInAccount> value) {
+    if (value is Future<GoogleSignInAccount>) {
```
Comment (pulyaevskiy) [inline packages/google_sign_in/lib/google_sign_in.dart]: Added around `else`.

**Comment by mehmetf** [comment]
This looks like it will fix https://github.com/flutter/flutter/issues/10552 as well.

**Comment by tvolkert** [comment]
Changes LGTM to proceed to documentation.

File: packages/google_sign_in/lib/google_sign_in.dart [hunk 8]
```diff
@@ -126,31 +119,80 @@ class GoogleSignIn {
...
+  /// Authentication can be re-initiated only after [signOut] or [disconnect].
+  Future<GoogleSignInAccount> signInSilently() =>
+      _addMethodCall('signInSilently');
 
   /// Starts the sign-in process.
   ///
-  /// If there is already a signed-in user, returns it.
-  Future<GoogleSignInAccount> signIn() async => _currentUser ?? _callMethod('signIn');
+  /// Interactive sign-in flow is triggered only be the first call to this method,
+  /// any consequent calls resolve to the same user instance or `null`. When
```
Comment (goderbauer) [inline packages/google_sign_in/lib/google_sign_in.dart]: Is this correct? If I abort the first call to `signIn` and then when `signIn` is called again it just returns `null` instead of re-triggering the sign in process? I would expect `signIn` to re-trigger the sign in process if no valid signed-in users is available.

File: packages/google_sign_in/lib/google_sign_in.dart [hunk 9]
```diff
@@ -126,31 +119,80 @@ class GoogleSignIn {
...
+  /// Keeps track of the most recently scheduled method call.
+  _MethodCompleter _lastMethodCompleter;
+
+  /// Adds call to [method] in a queue for execution.
+  ///
+  /// At most one in flight call is allowed to prevent concurrent (out of order)
+  /// updates to [currentUser] and [onCurrentUserChanged].
+  Future<GoogleSignInAccount> _addMethodCall(String method) {
+    if (_lastMethodCompleter == null) {
+      _lastMethodCompleter = new _MethodCompleter(method);
```
Comment (goderbauer) [inline packages/google_sign_in/lib/google_sign_in.dart]: nit: a cascade would look good here: `new _MethodCompleter(method)..complete(_callMethod(method))`

File: packages/google_sign_in/lib/google_sign_in.dart (hunk 8 above)
Comment (mehmetf) [inline packages/google_sign_in/lib/google_sign_in.dart]: We should definitely allow retrying signIn. It can fail/abort for many reasons.

Comment (pulyaevskiy) [inline packages/google_sign_in/lib/google_sign_in.dart]: You are right, this is not entirely correct. If previous call resolves to `null` the next call will re-trigger sign in process.

However, this just made me realize it would not be the case for calls within single iteration of event loop. Those return the same Future because of the if statement above.

I feel like I should probably get rid of that if statement after all. This way if the first call resolves to a user all subsequent calls will resolve to the same user. And if the first call resolves to `null` then the next call will re-trigger sign in flow, and all the calls after it will follow the same path.

Does it make sense?

I could also add similar handling for sign out methods so that if current user is already `null` then just complete immediately without triggering native code.

Comment (pulyaevskiy) [inline packages/google_sign_in/lib/google_sign_in.dart]: I realized I can't use the same pattern for "sign out" methods because `signOut` and `disconnect` perform different operations.

If a user was signed out by a call to `signOut` and we get a `disconnect` call we still need to call native side to revoke connection.

**Comment by pulyaevskiy** [comment]
Just pushed one more update with better handling of re-authentication. Also tried my best with the docs.

One thing I realized but I'm not sure if it's in the scope of this PR: `signOut` and `disconnect` signatures say they return `Future<GoogleSignInAccount>` which is not entirely true because it's actually always `Future<Null>`.

**Comment by goderbauer** [comment]
> One thing I realized but I'm not sure if it's in the scope of this PR: signOut and disconnect signatures say they return Future<GoogleSignInAccount> which is not entirely true because it's actually always Future<Null>.

You are right, we should probably change that. That can be in a different PR, though.

**Comment by tvolkert** [comment]
@pulyaevskiy can you bump the `CHANGELOG.md` and `pubspec.yaml`?  Thanks for the contribution!

**Comment by goderbauer** [comment]
@pulyaevskiy Your change is live: https://pub.dartlang.org/packages/google_sign_in

Thanks again for contributing!

**Comment by pulyaevskiy** [comment]
This is awesome, thanks everyone!
